UOP1_Project/Assets/Scripts/Pool/Pool.cs

```diff
@@ -0,0 +1,63 @@
+﻿using OP1.Factory;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace OP1.Pool
+{
+	/// <summary>
+	/// A generic pool that generates members of type T on-demand via a factory.
+	/// </summary>
+	/// <typeparam name="T">Specifies the type of elements to pool.</typeparam>
+	public abstract class Pool<T> : ScriptableObject, IPool<T> where T : IPoolable
+	{
+		protected readonly Stack<T> _available = new Stack<T>();
+		public abstract IFactory<T> Factory { get; set; }
+
+		public virtual T Add()
+		{
+			return Factory.Create();
+		}
+
+		public virtual T Request()
+		{
+			if (_available.Count <= 0)
+			{
+				_available.Push(Add());
+			}
+			T member = _available.Pop();
+			member.Initialize();
+			return member;
+		}
+
+		public virtual IEnumerable<T> Request(int num = 1)
+		{
+			List<T> members = new List<T>();
```

Comment:
When you know the number of elements, use either an array, or the `List` overload that takes in the capacity. 
``` suggestion
			List<T> members = new List<T>(num);
``` 
Depending on the size, the performance hit can be quite noticeable.